### PR TITLE
Implement updater for Windows and Linux

### LINKED ISSUE #94: Windows updater
### Role

I play minecraft in general

### Suggestion

Continue developing the windows updater that was being develop in poymcl

### Benefit

It would be easier to update polymc

### This suggestion is unique

- [X] I have searched the issue tracker and did not find an issue describing my suggestion, especially not one that has been rejected.

### You may use the editor below to elaborate further.

_No response_

### LINKED ISSUE #167: improve windows support
tracking issue for improving windows support in general

- [x] making things work fine on windows 7 again (#134)
- [x] adding updater (#94)
- [x] making windows 64 bit builds work
- [x] msvc builds

### PR DESCRIPTION
**Pull request by Ryex**
branch rename of #1097

Requires #1102 and #1227

Close #94 
Close #167

As I previously discussed on the discord the options for auto updating on windows are abysmal. we are frankly better off rolling our own (simple) solution.

## The Plan

- [x] Pull release info from the github api
- [x] Check if there is a newer release
- [x] Ask the user if they wish to update
- [x] Launch update process and exit prism.
- [x] (Optional) Let the user select from available newer releases
- [x] Download appropriate release asset to temp location (accounting from architecture, portable install etc. Simple so long as our naming remains consistent) (make the algo fuzy to account for variance if it's is needed)
- [x] create local file that marks an update in progress
- [x] if there are files we **know** need to be removed between updates "remove" them but create back-up (os move)
- [x] perform update
  - [x] IF PORTABLE: 
    - [x] Extract to temp local folder
    - [x] create maker file in tmp folder so updater knows what's going on (contains target install path)
    - [x] Run new updater, exit current updater process
    - [x] new updater process moves files out of temp dir to install dir and copies itself over old updater then exits 
  - [x] IF MSI / Setup.exe INSTALL:
    - [x] Run installer and exit updater
  - [x] IF Appimage 
    - [x] attempt to call embedded `AppImageUpdate`
- [x] Next time prism starts is sees marker files, present user with update log and cleans up.

Presumably at every point in the process if there is an error there is still an intact launcher install

### CODE REVIEW

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -275,12 +283,12 @@ jobs:
         if: runner.os == 'Windows' && matrix.msystem != ''
         shell: msys2 {0}
         run: |
-          cmake -S . -B ${{ env.BUILD_DIR }} -DCMAKE_INSTALL_PREFIX=${{ env.INSTALL_DIR }} -DCMAKE_BUILD_TYPE=${{ inputs.build_type }} -DENABLE_LTO=ON -DLauncher_BUILD_PLATFORM=${{ matrix.name }} -DCMAKE_C_COMPILER_LAUNCHER=${{ env.CCACHE_VAR }} -DCMAKE_CXX_COMPILER_LAUNCHER=${{ env.CCACHE_VAR }} -DLauncher_QT_VERSION_MAJOR=6 -DCMAKE_OBJDUMP=/mingw64/bin/objdump.exe -G Ninja
+          cmake -S . -B ${{ env.BUILD_DIR }} -DCMAKE_INSTALL_PREFIX=${{ env.INSTALL_DIR }} -DCMAKE_BUILD_TYPE=${{ inputs.build_type }} -DENABLE_LTO=ON -DLauncher_BUILD_PLATFORM=${{ matrix.name }} -DCMAKE_C_COMPILER_LAUNCHER=${{ env.CCACHE_VAR }} -DCMAKE_CXX_COMPILER_LAUNCHER=${{ env.CCACHE_VAR }} -DLauncher_QT_VERSION_MAJOR=6 -DCMAKE_OBJDUMP=/mingw64/bin/objdump.exe -DLauncher_UPDATER_GITHUB_REPO=https://github.com/${{ github.repository }} -G Ninja
```
Comment (Scrumplex) [inline .github/workflows/build.yml]: let's move this into our CMakeLists.txt

Everything else (i.e. macOS updater) is also in there anyway

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -275,12 +283,12 @@ jobs:
...
 
       - name: Configure CMake (Windows MSVC)
         if: runner.os == 'Windows' && matrix.msystem == ''
         run: |
-          cmake -S . -B ${{ env.BUILD_DIR }} -DCMAKE_INSTALL_PREFIX=${{ env.INSTALL_DIR }} -DCMAKE_BUILD_TYPE=${{ inputs.build_type }} -DENABLE_LTO=ON -DLauncher_BUILD_PLATFORM=${{ matrix.name }} -DLauncher_QT_VERSION_MAJOR=${{ matrix.qt_ver }} -DCMAKE_MSVC_RUNTIME_LIBRARY="MultiThreadedDLL" -A${{ matrix.architecture}} -DLauncher_FORCE_BUNDLED_LIBS=ON
+          cmake -S . -B ${{ env.BUILD_DIR }} -DCMAKE_INSTALL_PREFIX=${{ env.INSTALL_DIR }} -DCMAKE_BUILD_TYPE=${{ inputs.build_type }} -DENABLE_LTO=ON -DLauncher_BUILD_PLATFORM=${{ matrix.name }} -DLauncher_QT_VERSION_MAJOR=${{ matrix.qt_ver }} -DCMAKE_MSVC_RUNTIME_LIBRARY="MultiThreadedDLL" -A${{ matrix.architecture}} -DLauncher_FORCE_BUNDLED_LIBS=ON -DLauncher_UPDATER_GITHUB_REPO=https://github.com/${{ github.repository }}
```
Comment (Scrumplex) [inline .github/workflows/build.yml]: ditto

File: .github/workflows/build.yml [hunk 3]
```diff
@@ -295,7 +303,7 @@ jobs:
       - name: Configure CMake (Linux)
         if: runner.os == 'Linux'
         run: |
-          cmake -S . -B ${{ env.BUILD_DIR }} -DCMAKE_INSTALL_PREFIX=/usr -DCMAKE_BUILD_TYPE=${{ inputs.build_type }} -DENABLE_LTO=ON -DLauncher_BUILD_PLATFORM=Linux -DCMAKE_C_COMPILER_LAUNCHER=${{ env.CCACHE_VAR }} -DCMAKE_CXX_COMPILER_LAUNCHER=${{ env.CCACHE_VAR }} -DLauncher_QT_VERSION_MAJOR=${{ matrix.qt_ver }} -G Ninja
+          cmake -S . -B ${{ env.BUILD_DIR }} -DCMAKE_INSTALL_PREFIX=/usr -DCMAKE_BUILD_TYPE=${{ inputs.build_type }} -DENABLE_LTO=ON -DLauncher_BUILD_PLATFORM=Linux -DCMAKE_C_COMPILER_LAUNCHER=${{ env.CCACHE_VAR }} -DCMAKE_CXX_COMPILER_LAUNCHER=${{ env.CCACHE_VAR }} -DLauncher_QT_VERSION_MAJOR=${{ matrix.qt_ver }} -DLauncher_UPDATER_GITHUB_REPO=https://github.com/${{ github.repository }} -G Ninja
```
Comment (Scrumplex) [inline .github/workflows/build.yml]: ditto

File: .github/workflows/build.yml [hunk 4]
```diff
@@ -383,12 +391,13 @@ jobs:
         if: runner.os == 'Windows' && matrix.msystem == ''
         run: |
           cmake --install ${{ env.BUILD_DIR }} --config ${{ inputs.build_type }}
+          Get-Content ${{ env.BUILD_DIR }}/install_manifest.txt | %{ $_.TrimStart("$pwd/") } | %{ $_.TrimStart('${{ env.INSTALL_DIR }}') } | %{ $_.TrimStart('./') } | Out-File -Append -FilePath ${{ env.INSTALL_DIR }}/manifest.txt
```
Comment (DioEgizio) [inline .github/workflows/build.yml]: Why is it doing it twice???

**Review by DioEgizio — COMMENTED** [review]
Don't we also need to sign prismlauncher_updater too now?

File: .github/workflows/build.yml (hunk 4 above)
Comment (Ryex) [reply .github/workflows/build.yml]: merge artifact. I'll mark it for removal

**Comment by Ryex** [comment]
> Don't we also need to sign prismlauncher_updater too now?

we do, thanks for reminding me.

File: launcher/Application.cpp [hunk 5]
```diff
@@ -462,48 +567,48 @@ Application::Application(int &argc, char **argv) : QApplication(argc, argv)
...
+                 << qPrintable(QString(BuildConfig.LAUNCHER_COPYRIGHT).replace("\n", ", "));
         qDebug() << "Version                    : " << BuildConfig.printableVersionString();
         qDebug() << "Git commit                 : " << BuildConfig.GIT_COMMIT;
         qDebug() << "Git refspec                : " << BuildConfig.GIT_REFSPEC;
-        if (adjustedBy.size())
-        {
+        qDebug() << "Compiled for               : " << BuildConfig.systemID();
+        qDebug() << "Compiled by                : " << BuildConfig.compilerID();
+        qDebug() << "Build Artifact             : " << BuildConfig.BUILD_ARTIFACT;
+        qDebug() << "Updates Enabeled           : " << (updaterEnabled() ? "Yes" : "No");
```
Comment (DioEgizio) [inline launcher/Application.cpp]:
```suggestion
        qDebug() << "Updates Enabled           : " << (updaterEnabled() ? "Yes" : "No");
```

**Comment by Trial97** [comment]
~~Fixed the AppImageUpdate stuff, if I remember correctly windows update works.
So taking all this into consideration this should be ready for review, and maybe we organize a testing on the weekend for this.~~
Let me know if I missed anything. 
I did miss a lot of a stuff: 
- [x] appimageupdate is inside the appimage
- [x] and did not touch the portable stuff

**Comment by Trial97** [comment]
Now I'm pretty sure this is ready for review 
Working for me:
- linux portable update
- appimage update

I did not touch any windows related code so that one I expect it to run as before.

**Comment by DioEgizio** [comment]
> Now I'm pretty sure this is ready for review Working for me:
> 
>     * linux portable update
> 
>     * appimage update
> 
> 
> I did not touch any windows related code so that one I expect it to run as before.

i think we should have another test in #testing to see if everything works fine, just in case

**Review by DioEgizio — APPROVED** [review]

**Review by Trial97 — APPROVED** [review]

**Review by TayouVR — APPROVED** [review]

**Comment by RokeJulianLockhart** [comment]
This won't activate in managed versions like the Flatpak package, right?

**Comment by Trial97** [comment]
No Flatpack package should not be affected.

**Pull request merged**